lib/openhab/dsl/time/month_day.rb

```diff
@@ -52,54 +35,80 @@ def self.range?(range)
-            year = -> { Year.from(next_date) }
-            MonthDayRangeElement.new(month_day: MonthDay.from(next_date), year: year)
+            next_day_of_month = @month_day.day_of_month + 1
+            next_month = @month_day.month_value
+
+            if next_day_of_month > @month_day.month.max_length
+              next_day_of_month = 1
+              next_month = @month_day.month.plus(1).value
+            end
+
+            DayOfYear.new(month_day: MonthDay.of(next_month, next_day_of_month), range: @range)
+          end
+
+          #
+          # Offset by 1 year if the range begin is greater than the range end
+          # and if the month day is less than the begining of the range
+          # @return [Number] 366 if the month_day should be offset by a year
+          def offset
+            @range.begin > @range.end && month_day < @range.begin ? DAYS_IN_YEAR : 0
+          end
+
+          #
+          # Calculate the day within the range for the underlying month day
+          # @return [Number] Representation of the MonthDay as a number from 1 to 732
+          def day_in_range
+            @day_in_range ||= month_day.max_day_of_year + offset
           end
 
           # Compare MonthDayRangeElement to other objects as required by Range class
-          # rubocop:disable Metrics/AbcSize
-          # Case statement needs to work against multiple types
-          def <=>(other)
+          def <=>(other) # rubocop:disable Metrics/AbcSize
             case other
-            when LocalDate then to_local_date.compare_to(other)
-            when Date then self.<=>(LocalDate.of(other.year, other.month, other.day))
-            when MonthDay then self.<=>(MonthDayRange.current_year.call.at_month_day(other))
+            when DayOfYear then day_in_range.<=>(other.day_in_range)
+            when MonthDay then self.<=>(DayOfYear.new(month_day: other, range: @range))
+            when LocalDate then self.<=>(MonthDay.of(other.month_value, other.day_of_month))
+            when Date then self.<=>(MonthDay.of(other.month, other.day))
             else
-              return self.<=>(other.to_local_date) if other.respond_to? :to_local_date
-              return self.<=>(other.to_date) if other.respond_to? :to_date
+              return self.<=>(other.to_local_date) if other.respond_to?(:to_local_date)
+              return self.<=>(other.to_date) if other.respond_to?(:to_date)
 
               raise "Unable to convert #{other.class} to compare to MonthDay"
             end
           end
-          # rubocop:enable Metrics/AbcSize
+        end
+      end
+
+      java_import java.time.Month
+      # Extend Month with helper method
+      class Month
+        # Calcalute and memoize the maximum number of days in a year before this month
+        # @return [Number] maximum nummber of days in a year before this month
+        def max_days_before
+          @max_days_before ||= Month.values.select { |month| month < self }.sum(&:max_length)
```

Comment:
This would save us from running the select loop, perhaps at the expense of readability - your call
```suggestion
          @max_days_before ||= Month.values.sum(&:max_length) - max_length
```